datapackage_pipelines/lib/unpivot.py

```diff
@@ -0,0 +1,100 @@
+import copy
+import re
+
+from datapackage_pipelines.wrapper import ingest, spew
+from datapackage_pipelines.utilities.resource_matcher import ResourceMatcher
+
+parameters, datapackage, resource_iterator = ingest()
+
+resources = ResourceMatcher(parameters.get('resources'))
+unpivot_fields = parameters.get('unpivot')
+keys_ = parameters.get('extraKeyFields')
+values_ = parameters.get('extraValueField')
+unpivot_fields_without_regex = []
+
+
+def match_fields(field_name_re, expected):
+    def filt(field):
+        return (field_name_re.fullmatch(field['name']) is not None) is expected
+    return filt
+
+
+def process_datapackage(datapackage_):
+    for resource in datapackage_['resources']:
+        name = resource['name']
+        if not resources.match(name):
+            continue
+
+        if 'schema' not in resource:
+            continue
+
+        fields = resource.setdefault('schema', {}).get('fields', [])
+
+        for u_field in unpivot_fields:
+            field_name_re = re.compile(u_field['name'])
+            fields_to_pivot = (list(
+                filter(match_fields(field_name_re, True), fields)
+            ))
+            fields = list(
+                filter(match_fields(field_name_re, False), fields)
+            )
+
+            # handle with regex
+            for field_to_pivot in fields_to_pivot:
+                oiginal_key_values = u_field['keys']  # With regex
+                new_key_values = {}
+                for key in oiginal_key_values:
+                    new_val = re.sub(
```

Comment:
Why do we need `parse_field` here?
If it is a string, then we use regexp (and get a string in return).
If it's not, we just copy the value (as obviously it's not a regexp).

Reply:
@akariv this is the case when it is a string, we use regex, but not want a string in return:
* imagine this is the title of the original field: `incomes 2017` 
* and we want to grab 2017 with regex `\2` from `(\w+) ([0-9]{4})`
* result will be "2017" as string
* but user may need it as number or float etc.. depending type defined in `extraKeyFields`

Reply:
you can let dpp take care of type conversion, and leave it as a string here.